kubernetes/client.go

```diff
@@ -123,34 +125,128 @@ func (in *IstioClient) GetServices(namespaceName string) (*v1.ServiceList, error
 // A service is defined by the namespace and the service name.
 // It returns an error on any problem.
 func (in *IstioClient) GetServiceDetails(namespace string, serviceName string) (*ServiceDetails, error) {
-	service, err := in.k8s.CoreV1().Services(namespace).Get(serviceName, emptyGetOptions)
-	if err != nil {
-		return nil, err
+	serviceChan, endpointsChan := make(chan *v1.Service), make(chan *v1.Endpoints)
+	errChan := make(chan error)
+
+	go func() {
```

Comment:
I really like the usage of go routines for better performance on http calls. However, IMO, I would move this anonymous functions out of this main function. This GetServiceDetails already has 80 LOC, and I think, a best practice would be to split it down. In addition, service seems to be one of the most important subjects in this repo, therefore this method probably grow.

Reply:
I understand your comment, but I have a mixed feeling.
I basically followed the same pattern introduced in
https://github.com/swift-sunshine/swscore/blob/master/prometheus/metrics.go#L161
where anonymous functions were used in a similar way.

Would it be ok if we split the client in the methods specific for kubernetes and istio separately ?
But I maintain the anonymous functions in the same way (as it is just a statement call and a validation).
Should that address your comment ?

Reply:
I was thinking that It would be great to distribute the code a little bit.
What about something like this?

```go
func (in *IstioClient) GetServiceDetails(namespace string, serviceName string) (*ServiceDetails, error) {
	var serviceDetails = ServiceDetails{name: serviceName, namespace: namespace}
	serviceChan, errChan := serviceDetails.paralelFetchService()
	endpointsChan, errChan := serviceDetails.paralelFetchEndpoints()
	podChan, podErrChan := serviceDetails.paralelFetchPods()

	if err := serviceDetails.harvestServiceDetails(serviceChan, errChan); err != nil {
		return nil, err
	}

	if err := serviceDetails.harvestEndpoints(endpointsChan, errChan); err != nil {
		return nil, err
	}

	if err := serviceDetails.harvestPods(podChan, podErrChan); err != nil {
		return nil, err
	}

	return &serviceDetails, nil
}
```

and have another file in charge of take the result from the API and load it into the service details fields:

```go
package kubernetes

func (sd *ServiceDetails) fetchService() (<-chan *v1.Service, <-chan error) {
	serviceChan := make(chan *v1.Service)
	errChan := make(chan error)

	go func() {
		service, err := in.k8s.CoreV1().Services(sd.Namespace).Get(sd.Name, emptyGetOptions)
		if err != nil {
			errChan <- err
		} else {
			serviceChan <- service
		}
	}()

	return serviceChan, errChan
}

func (sd *ServiceDetails) harvestServiceDetails(serviceChan <-chan *v1.Service, errChan <-chan error) error {
	select {
	case sd.Service = <-serviceChan:
	case serviceDetailsError = <-errChan:
	}

	return serviceDetailsError
}

func (sd *ServiceDetails) getPodNames() ([]string, error) {}

func (sd *ServiceDetails) fetchPods() (<-chan []*v1.Pod, <-chan error) {}

func (sd *ServiceDetails) harvestPodsfrom(<-chan []*v1.Pod, <-chan error) error {}
```

I might be a little bit picky (tell me if so, with no problems). Show me your thoughts on that :)

Reply:
>  Would it be ok if we split the client in the methods specific for kubernetes and istio separately ?
I do often like split things, hehe. If you feel that now is the time to do it, I totaly support. Otherwise, we can do it in separete PR/JIRA.